### PR TITLE
WIP: Support NCHW (for conv2d).

### PR DESCRIPTION
**Pull request by TimZaman**
Tensorflow (at least versions 1.3 and below) does not support `NCHW` computations of many ops when run on the CPU. That creates compatibility issues. These are originally mitigated in Keras, by doing a "NHWC roundtrip" and any op requested to be ran in "`channels_first`" (`NCHW`) would actually be transposed to- and from-  `NHWC`.

`NCHW` often has performance benefits; for example, cuDNN is often fastest when the largest dimension is last (favouring `NCHW` for images).

The current implementation probes if NCHW is available by checking:
1. Whether `NCHW` (`channels_first`) is requested at all
2. If `NCHW` is supported by checking (1) if a GPU is available at all and (2) if it is not explicitly set on CPU.

If the op is set explicitly on the CPU, or there is no GPU device, and if NCHW is requested, we will proceed with the transpose roundtrip as we always did in Keras.

This PR only patches the `conv2d` op. There are more (and more significant) ops to be patched: any op that uses `_postprocess_conv2d_output` or `_postprocess_conv3d_output` ad `bias_add` too. The latter is actually Keras's bottleneck when it comes to NCHW. Lets save those for distinct PRs.

### CODE REVIEW

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -193,6 +201,64 @@ def set_session(session):
...
+        """__init__ method."""
+        self.device = None
+
+    def _set_device(self, device):
+        """_set_device method."""
+        self.device = device
+
+
+def get_current_tf_device():
+    """Return device string of current graph context that's explicitly, otherwise returns `None`."""
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Please fix docstring typos

File: keras/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -193,6 +201,64 @@ def set_session(session):
...
+    def __init__(self):
+        """__init__ method."""
+        self.device = None
+
+    def _set_device(self, device):
+        """_set_device method."""
+        self.device = device
+
+
+def get_current_tf_device():
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: I think we should make this function private, as well as `is_current_explicit_device`, and `get_available_gpus`.

File: keras/backend/tensorflow_backend.py [hunk 3]
```diff
@@ -193,6 +201,64 @@ def set_session(session):
...
+    """Return device string of current graph context that's explicitly, otherwise returns `None`."""
+    g = tf.get_default_graph()
+    op = _TfDeviceCaptureOp()
+    g._apply_device_functions(op)
+    return op.device
+
+
+def is_current_explicit_device(device_type):
+    """
+    Check if the current device is explicitly set on the device type.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Please put the docstring description on the first line

File: keras/backend/tensorflow_backend.py [hunk 4]
```diff
@@ -193,6 +201,64 @@ def set_session(session):
...
+    device = get_current_tf_device()
+    return (device is not None and device.device_type == device_type.upper())
+
+
+def get_available_gpus():
+    """Get a list of available gpu devices (formatted as strings)."""
+    return [x.name for x in LOCAL_DEVICES if x.device_type == 'GPU']
+
+
+def has_nchw_support():
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: This sounds too specific. Why not something like "_running_on_gpu()" or something to that extent?

**Comment by Dref360** [comment]
Could we have some tests where we mock being on a gpu machine so that some tests run using NHWC and some run using NCWH? We could test `with tf.device('/cpu:0'):` etc.

File: keras/backend/tensorflow_backend.py (hunk 4 above)
Comment (TimZaman) [reply keras/backend/tensorflow_backend.py]: Hmm.. What it checks for is if the current scope is "not explicit on CPU, and has GPUs available". It's called like this to anticipate any TF 1.x's support that does not need the roundtrip so we can test for the tf version in the function.

**Comment by TimZaman** [comment]
Sure @Dref360 , that does require ignoring tests if people/travis does not have a GPU though. Do we already have tests in Keras that skip tests for environments where there's no GPU?

File: keras/backend/tensorflow_backend.py [hunk 5]
```diff
@@ -193,6 +201,72 @@ def set_session(session):
...
+        the device (`CPU` or `GPU`). If the scope is not explicitly set, it will
+        return `None`.
+    """
+    g = tf.get_default_graph()
+    op = _TfDeviceCaptureOp()
+    g._apply_device_functions(op)
+    return op.device
+
+
+def is_current_explicit_device(device_type):
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Please make this method private.

File: keras/backend/tensorflow_backend.py [hunk 6]
```diff
@@ -193,6 +201,72 @@ def set_session(session):
...
+        bool: if the current device scope is explicitly set on the device type.
+    """
+    device_type = device_type.upper()
+    if device_type not in ['CPU', 'GPU']:
+        raise ValueError('device_type should be either "CPU" or "GPU".')
+    device = _get_current_tf_device()
+    return (device is not None and device.device_type == device_type.upper())
+
+
+def get_available_gpus():
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Please make this method private (unless there is a rationale for making it part of the public API).

File: keras/backend/tensorflow_backend.py [hunk 7]
```diff
@@ -193,6 +201,72 @@ def set_session(session):
...
+    """Get a list of available gpu devices (formatted as strings).
+
+    # Returns
+        A list of available GPU devices.
+    """
+    return [x.name for x in LOCAL_DEVICES if x.device_type == 'GPU']
+
+
+def has_nchw_support():
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Please make this method private.

File: keras/backend/tensorflow_backend.py [hunk 8]
```diff
@@ -44,6 +45,13 @@
 # Change its value via `manual_variable_initialization(value)`.
 _MANUAL_VAR_INIT = False
 
+# This map is for converting the keras data format string to that of TF.
+DATA_FORMAT_MAP = {'channels_first': 'NCHW', 'channels_last': 'NHWC'}
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Please make this global variable private.

File: keras/backend/tensorflow_backend.py [hunk 9]
```diff
@@ -44,6 +45,13 @@
...
+
+# This list queries the devices.
+# We assume our devices don't change during our lifetime.
+LOCAL_DEVICES = device_lib.list_local_devices()
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Please make this global variable private.

**Comment by fchollet** [comment]
Please fix the docstrings and PEP8 issue reported in CI: https://travis-ci.org/fchollet/keras/builds/282490074

File: keras/backend/tensorflow_backend.py [hunk 10]
```diff
@@ -193,6 +201,72 @@ def set_session(session):
...
+
+def _is_current_explicit_device(device_type):
+    """Check if the current device is explicitly set on the device type specified.
+
+    # Arguments
+        device_type: A string containing `GPU` or `CPU` (case-insensitive).
+
+    # Returns
+        bool: if the current device scope is explicitly set on the device type.
+    """
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Per the failing docstring test, this docstring needs a `Raises` section mentioning the ValueError: https://travis-ci.org/fchollet/keras/jobs/282558708

**Comment by TimZaman** [comment]
OK! We're not done yet; we need to patch other ops too, most importantly (biggest bottleneck) is the add_bias's tranpose on NCHW. I'll get to that.

**Pull request merged**